10 Notes/Component Specification - Prover.md

```diff
@@ -0,0 +1,388 @@
+| `KeyedCompression`     | Keyed 2->1 compression where $key \in \{0,1,2,3\}$ .                                                    | -                                    | `key`, `inp[2]`      | `out`                         |
+| `ExtractLowerBits`     | Extracts the lower `n` bits of `inp` (LSB‑first).                                                       | `n`                                  | `inp`                | `out[n]`                      |
+| `Log2`                 | Checks `inp == 2^out` with `0 < out <= n`. Also emits a mask vector with ones for indices `< out`.      | `n`                                  | `inp`                | `out`, `mask[n+1]`            |
+| `CeilingLog2`          | Computes `ceil(log2(inp))` and returns the bit‑decomposition and a mask.                                | `n`                                  | `inp`                | `out`, `bits[n]`, `mask[n+1]` |
+| `BinaryCompare`        | Compares two n‑bit numbers `A` and `B` (LSB‑first); outputs `-1` if `A<B`, `0` if equal, `+1` if `A>B`. | `n`                                  | `A[n]`, `B[n]`       | `out`                         |
+
+## 3. Functional Requirements
+
+**Data Commitment**
+- Fetch existing slot commitments using `BlockStore` and `SlotsBuilder`:  `cell -> block -> slot` Merkle trees for each slot in the locally stored dataset. 
+- Fetch dataset commitment: `slot -> dataset` verification tree root.  
+- Proof material: retrieve cell data (as field elements).
+
+**Sampling**
+- Checks for marketplace challenges per slot. 
+- Random sampling: Derive `nSamples` cell indices for the `slotIndex` from `(entropy, slotRoot)`. 
+- For each sampled cell, fetch: `cellData` (as field elements) and Merkle paths (all `cell -> block -> slot -> dataset`)
+- Generate `ProofInputs` containing the public inputs (`datasetRoot`, `slotIndex`, `entropy`) and private witness (`cellData`, `slotRoot`, `MerklePaths`).
+
+**Proof Generation**
+- Given `ProofInputs`, use the configured backend (Groth16 over BN254) to create a succinct Groth16 proof. 
+- The circuit enforces the same Merkle layout and Poseidon2 hashing used for commitments
+
+## 4. Non-Functional Requirements
+
+**Performance / Latency**
+- End-to-end (sample -> prove -> submit) completes within the on-chain proof window with some safety margin. 
+- Small Proof size, e.g. Groth16/BN254 plus public inputs. 
+- On-chain verification cost is minimal.
+- Support concurrent proving for multiple slots.
+
+**Security & Correctness**
+- Soundness and completeness: only accept valid proofs, invalid inputs must not yield accepted proofs. 
+- Commitment integrity: proofs are checked against the publicly available Merkle commitments to the stored data.
+- Entropy binding: sample indices must be derived from the on-chain entropy and the slot commitment. 
```

Comment:
I think these requirements are good and most of them are pretty obvious why they are important. But for this last one I'd suggest adding why entropy is needed. If I'm not mistaken, it's so that the required proofs can't be predicted and calculated ahead of time.